### PR TITLE
Fix GetMiningTimestamp

### PR DESCRIPTION
**Pull request by quantumagi**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs [hunk 1]
```diff
@@ -45,38 +48,64 @@ public SlotsManager(Network network, IFederationManager federationManager, Chain
...
+            uint lastRoundStart = tip.Header.Time - (uint)(lastMinerIndex * roundTime / federationMembers.Count);
+
+            // Determine the slot in which we're mining.
+            int thisMinerIndex = federationMembers.FindIndex(m => m.PubKey == this.federationManager.CurrentFederationKey.PubKey);
 
-            // Time when current round started.
-            uint roundStartTimestamp = (currentTime / roundTime) * roundTime;
-            uint nextTimestampForMining = roundStartTimestamp + slotIndex * this.consensusOptions.TargetSpacingSeconds;
+            // To calculate the latest round start bring the last round forward in round time increments but stop short of the current time.
+            uint prevRoundStart = lastRoundStart + ((currentTime - lastRoundStart) / roundTime) * roundTime;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs:75]: Please check this calculation as the results will always be `currentTime `.

File: src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs [hunk 2]
```diff
@@ -45,38 +48,64 @@ public SlotsManager(Network network, IFederationManager federationManager, Chain
...
 
             return nextTimestampForMining;
         }
 
+
         /// <inheritdoc />
         public bool IsValidTimestamp(uint headerUnixTimestamp)
         {
-            return (headerUnixTimestamp % this.consensusOptions.TargetSpacingSeconds) == 0;
+            return true;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs:108]: Can we remove this method then?

**Comment by fassadlr** [comment]
@quantumagi those integration tests are failing cause the changes to slotsmanager doesnt take into account when tip is genesis.

**Comment by fassadlr** [comment]
```
            IFederationMember lastMiner = this.federationHistory.GetFederationMemberForBlock(tip, federationMembers);
            int lastMinerIndex = federationMembers.FindIndex(m => m.PubKey == lastMiner.PubKey);
```

`lastMiner` is null here.

File: src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs [hunk 3]
```diff
@@ -45,40 +45,57 @@ public SlotsManager(Network network, IFederationManager federationManager, Chain
...
+            ChainedHeader tip = this.chainIndexer.Tip;
+            List<IFederationMember> federationMembers = this.federationHistory.GetFederationForBlock(tip, true);
 
             // Round length in seconds.
             uint roundTime = this.GetRoundLengthSeconds(federationMembers.Count);
 
-            // Index of a slot that current node can take in each round.
-            uint slotIndex = (uint)federationMembers.FindIndex(x => x.PubKey == this.federationManager.CurrentFederationKey.PubKey);
+            // Determine when the last round started by looking at who mined the tip.
+            IFederationMember lastMiner = this.federationHistory.GetFederationMemberForBlock(tip, federationMembers);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs:64]: This fixed the tests:

```
            IFederationMember lastMiner = null;
            if (tip.Height > 0)
                lastMiner = this.federationHistory.GetFederationMemberForBlock(tip, federationMembers);
            else
                lastMiner = this.federationManager.GetCurrentFederationMember();

```

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs:64]: If this is correct, can you commit?

**Comment by quantumagi** [comment]
I've changed `GetFederationMemberForBlock` to return the last federation member.

**Review by fassadlr — APPROVED** [review]

**Pull request merged**